addons/controllers/utils/utils.go

```diff
@@ -11,28 +11,80 @@ import (
 )
 
-// GetVSphereCluster gets the VSphereCluster CR for the cluster object
-func GetVSphereCluster(ctx context.Context, clt client.Client, cluster *clusterapiv1beta1.Cluster) (*capvv1beta1.VSphereCluster, error) {
-	vsphereCluster := &capvv1beta1.VSphereCluster{}
-	if err := clt.Get(ctx, types.NamespacedName{
-		Namespace: cluster.Namespace,
-		Name:      cluster.Name,
-	}, vsphereCluster); err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil, errors.Errorf("VSphereCluster %s/%s not found", cluster.Namespace, cluster.Name)
-		}
-		return nil, errors.Errorf("VSphereCluster %s/%s could not be fetched, error %v", cluster.Namespace, cluster.Name, err)
+// GetVSphereClusterParavirtual gets the VSphereCluster CR for the cluster object in paravirtual mode
+func GetVSphereClusterParavirtual(ctx context.Context, clt client.Client, cluster *clusterapiv1beta1.Cluster) (*capvvmwarev1beta1.VSphereCluster, error) {
+
+	vsphereClusters := &capvvmwarev1beta1.VSphereClusterList{}
+	labelMatch, err := labels.NewRequirement(clusterapiv1beta1.ClusterLabelName, selection.Equals, []string{cluster.Name})
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector()
+	labelSelector = labelSelector.Add(*labelMatch)
+	if err := clt.List(ctx, vsphereClusters, &client.ListOptions{LabelSelector: labelSelector, Namespace: cluster.Namespace}); err != nil {
+		return nil, err
+	}
+	if len(vsphereClusters.Items) != 1 {
+		return nil, fmt.Errorf("expected to find 1 VSphereCluster object for label key %s and value %s but found %d",
+			clusterapiv1beta1.ClusterLabelName, cluster.Name, len(vsphereClusters.Items))
+	}
+	return &vsphereClusters.Items[0], nil
+}
+
+// GetVSphereClusterNonParavirtual gets the VSphereCluster CR for the cluster object in non-paravirtual mode
+func GetVSphereClusterNonParavirtual(ctx context.Context, clt client.Client, cluster *clusterapiv1beta1.Cluster) (*capvv1beta1.VSphereCluster, error) {
+
+	vsphereClusters := &capvv1beta1.VSphereClusterList{}
+	labelMatch, err := labels.NewRequirement(clusterapiv1beta1.ClusterLabelName, selection.Equals, []string{cluster.Name})
+	if err != nil {
+		return nil, err
+	}
+	labelSelector := labels.NewSelector()
+	labelSelector = labelSelector.Add(*labelMatch)
+	if err := clt.List(ctx, vsphereClusters, &client.ListOptions{LabelSelector: labelSelector, Namespace: cluster.Namespace}); err != nil {
+		return nil, err
+	}
+	if len(vsphereClusters.Items) != 1 {
+		return nil, fmt.Errorf("expected to find 1 VSphereCluster object for label key %s and value %s but found %d",
+			clusterapiv1beta1.ClusterLabelName, cluster.Name, len(vsphereClusters.Items))
+	}
+	return &vsphereClusters.Items[0], nil
+}
+
+// GetControlPlaneVsphereMachineTemplate gets the VsphereMachineTemplate CR of control plane
+func GetControlPlaneVsphereMachineTemplate(ctx context.Context, clt client.Client, cluster *clusterapiv1beta1.Cluster) (*capvv1beta1.VSphereMachineTemplate, error) {
```

Comment:
ControlPlaneVsphereMachineTemplateForCluster would be a more accurate function name (dropping Get is more conventional in k8s/capi).